### DISCUSSION TITLE
Question about disabling plugin/extension

### DISCUSSION THREAD

**Discussion by zsan** [rubycdp/ferrum#337] [General]
Hello,

Just for my curiosity, is that possible to disable plugin with Ferrum or CDP ?

For example, i have this code with capybara to disable PDF Viewer 

```
Capybara.register_driver :chrome do |app|
  Capybara::Selenium::Driver.load_selenium

  browser_options = Selenium::WebDriver::Chrome::Options.new

  browser_options.add_preference(
    :download,
    directory_upgrade: true,
    prompt_for_download: false,
    extensions_to_open: "applications/pdf"    
  )

  browser_options.add_preference(
    :plugins,
    always_open_pdf_externally: true,
    plugins_disabled: ["Chrome PDF Viewer"]    
  )

  Capybara::Selenium::Driver.new(app, browser: :chrome, options: browser_options)
end
```

**Comment by route** [comment]
If you have a question, please open discussion, not an issue

**Reply by route** [reply]
For plugins you mean extensions? Extensions don't work in headless Chrome. https://peter.sh/experiments/chromium-command-line-switches/#disable-extensions-except

**Reply by zsan** [reply]
I am not sure if its plugins or extension :=).

So let's say i have url `http://www.example.com/invoice.pdf` and when i get the page with ferrum then i'll see the browser open the pdf with browser, but i do not want that behaviour, i want browser download the pdf file into directory that is defined in `save_path`.

But you're correct, in `headless` mode, the pdf file is automatically downloaded, and i am wondering if there is a way to imitate that in `non-headless` mode ?

I tried with this setup and no luck

```ruby
browser = Ferrum::Browser.new(
  headless: false,
  save_path: "/tmp",
  timeout: 60,
  browser_options: {
    "disable-extensions": nil,
  },
)
```

**Reply by route** [reply]
This should have worked, what happens if you open like this?

**Reply by zsan** [reply]
> This should have worked, what happens if you open like this?

The PDF file is opened within a browser (not downloaded)

**Reply by route** [reply]
Does it have content disposition header?

**Reply by zsan** [reply]
> Does it have content disposition header?

I am not sure about that, how do i know if it have content disposition header ?

Or maybe you may want to try to real URL

```
browser = Ferrum::Browser.new(
  headless: false,
  save_path: "/tmp",
  timeout: 60,
  browser_options: {
    "disable-extensions": nil,
  },
)

browser.goto("https://www.africau.edu/images/default/sample.pdf")
```

**Reply by route** [reply]
Interesting I tried this one with curl and got application/pdf but with Chrome html. Could you try to make a request with headers like in curl and see if this worked?

**Reply by zsan** [reply]
Same thing, pdf is opened in chrome
Tried with this headers:

```
browser.headers.set({
  "Content-Type" => "application/pdf",
  "Accept" => "*.*",  
  "Accept-Ranges" => "bytes"
})
```

**Reply by route** [reply]
That's none of my business, but ok here you go:

```
require "ferrum"

browser = Ferrum::Browser.new(headless: false,  save_path: "/tmp",  timeout: 5)

browser.network.intercept(request_stage: 'Response')
browser.on(:request) do |request|
  if request.match?(/.*\.pdf\z/)
    request.respond(responseHeaders: request.headers.merge('content-disposition' => 'attachment'))
  else
    request.continue
  end
end

browser.goto("https://www.africau.edu/images/default/sample.pdf")
```

**Reply by zsan** [reply]
That's great, thank you
